src/coreclr/System.Private.CoreLib/src/System/Reflection/MethodBase.CoreCLR.cs

```diff
@@ -62,30 +65,59 @@ internal virtual Type[] GetParameterTypes()
+        private protected Span<object?> CheckArguments(ref StackAllocedArguments stackArgs, object?[]? parameters, Binder? binder,
             BindingFlags invokeAttr, CultureInfo? culture, Signature sig)
         {
-            // copy the arguments in a different array so we detach from any user changes
-            object[] copyOfParameters = new object[parameters.Length];
+            Debug.Assert(Unsafe.SizeOf<StackAllocedArguments>() == StackAllocedArguments.MaxStackAllocArgCount * Unsafe.SizeOf<object>(),
+                "MaxStackAllocArgCount not properly defined.");
 
-            ParameterInfo[]? p = null;
-            for (int i = 0; i < parameters.Length; i++)
+            Span<object?> copyOfParameters = default;
+
+            if (parameters is not null)
             {
-                object arg = parameters[i];
-                RuntimeType argRT = sig.Arguments[i];
+                // copy the arguments into a temporary buffer (or a new array) so we detach from any user changes
+                copyOfParameters = (parameters.Length <= StackAllocedArguments.MaxStackAllocArgCount)
+                        ? MemoryMarshal.CreateSpan(ref stackArgs._arg0, parameters.Length)
+                        : new Span<object?>(new object?[parameters.Length]);
 
-                if (arg == Type.Missing)
+                ParameterInfo[]? p = null;
+                for (int i = 0; i < parameters.Length; i++)
                 {
-                    p ??= GetParametersNoCopy();
-                    if (p[i].DefaultValue == System.DBNull.Value)
-                        throw new ArgumentException(SR.Arg_VarMissNull, nameof(parameters));
-                    arg = p[i].DefaultValue!;
+                    object? arg = parameters[i];
+                    RuntimeType argRT = sig.Arguments[i];
+
+                    if (arg == Type.Missing)
+                    {
+                        p ??= GetParametersNoCopy();
+                        if (p[i].DefaultValue == System.DBNull.Value)
+                            throw new ArgumentException(SR.Arg_VarMissNull, nameof(parameters));
+                        arg = p[i].DefaultValue!;
+                    }
+                    copyOfParameters[i] = argRT.CheckValue(arg, binder, culture, invokeAttr);
                 }
-                copyOfParameters[i] = argRT.CheckValue(arg, binder, culture, invokeAttr)!;
             }
 
             return copyOfParameters;
         }
+
+        // Helper struct to avoid intermediate object[] allocation in calls to the native reflection stack.
+        // Typical usage is to define a local of type default(StackAllocedArguments), then pass 'ref theLocal'
+        // as the first parameter to CheckArguments. CheckArguments will try to utilize storage within this
+        // struct instance if there's sufficient space; otherwise CheckArguments will allocate a temp array.
+        private protected struct StackAllocedArguments
```

Comment:
Q to reviewers: does this need `LayoutKind.Sequential`?

Reply:
StructLayout doesn't really do much with the managed view of non-blittable types. There was a loong discussion e.g. here: https://github.com/dotnet/runtime/issues/44579#issuecomment-725901040

Reply:
Ah, perfect! I missed that sentence in the doc.

> For non-blittable types, it controls the layout when the class or structure is marshaled to unmanaged code, but does not control the layout in managed memory.